### PR TITLE
Add support for AArch64 SHA extensions.

### PR DESCRIPTION
**Pull request by jasondavies**

### CODE REVIEW

**Comment by jasondavies** [comment]
OK, now a new error: `LLVM ERROR: Cannot select: intrinsic %llvm.arm.neon.sha1h`.

This is when running `TARGET=aarch64-unknown-linux-gnu cargo test`.

**Comment by jasondavies** [comment]
Oh wait, I probably need to set target_feature `crypto`.

**Comment by gnzlbg** [comment]
Try with `RUSTFLAGS="-C target-feature=+neon"`, depending on the ARM target you are testing (arm vs armv7 vs armv8 vs aarch64vs.. ) you might need to enable a "baseline" feature to get these to work.

File: coresimd/aarch64/neon.rs [hunk 1]
```diff
@@ -138,6 +138,27 @@ extern "C" {
     #[link_name = "llvm.aarch64.neon.fminv.f64.v2f64"]
     fn vminvq_f64_(a: float64x2_t) -> f64;
 
+    #[link_name = "llvm.arm.neon.sha1h"]
```
Comment (gnzlbg) [inline coresimd/aarch64/neon.rs]: llvm.arm.neon hints that they do belong in the `coresimd/arm/neon.rs` module

File: coresimd/aarch64/neon.rs [hunk 2]
```diff
@@ -380,6 +401,77 @@ pub unsafe fn vminvq_f64(a: float64x2_t) -> f64 {
     vminvq_f64_(a)
 }
 
+/// SHA1 hash update accelerator, choose.
+#[inline]
+#[target_feature(enable = "neon")]
```
Comment (gnzlbg) [inline coresimd/aarch64/neon.rs]: have you tried enabling "neon,crypto" here ?

Comment (gnzlbg) [reply coresimd/aarch64/neon.rs]: they probably just require the crypto feature

Comment (jasondavies) [reply coresimd/aarch64/neon.rs]: Yeah, (I did `#[target_feature(enable = "neon", enable = "crypto")]`).  No luck.

**Comment by jasondavies** [comment]
> `RUSTFLAGS="-C target-feature=+neon"`

Didn't help, unfortunately.

File: coresimd/aarch64/neon.rs (hunk 1 above)
Comment (jasondavies) [reply coresimd/aarch64/neon.rs]: OK, good! I wasn't sure if `crypto.rs` would make more sense.

**Comment by gnzlbg** [comment]
So I'll give these a try tomorrow (gotta sleep now), but what you can do in the meantime is check what clang does for these, in particular, clang / llvm tests, what features, or -mattr do they enable to compile these tests, and try those.

**Comment by jasondavies** [comment]
Turns out I should have been using `llvm.aarch64.crypto.sha1h`!  I guess it should be in `crypto.rs` after all.

File: coresimd/aarch64/crypto.rs [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+#[cfg_attr(test, assert_instr(sha1p))]
+pub unsafe fn vsha1pq_u32(hash_abcd: u32x4, hash_e: u32, wk: u32x4) -> u32x4 {
+    vsha1pq_u32_(hash_abcd, hash_e, wk)
+}
+
+/// SHA1 schedule update accelerator, first part.
+#[inline]
+#[target_feature(enable = "crypto")]
+#[cfg_attr(test, assert_instr(sha1su0))]
+pub unsafe fn vsha1su0q_u32(w0_3: u32x4, w4_7: u32x4, w8_11: u32x4) -> u32x4 {
```
Comment (alexcrichton) [inline coresimd/aarch64/crypto.rs]: I think these should use the `uint32x4_t` types nowadays, right @gnzlbg?

Comment (jasondavies) [reply coresimd/aarch64/crypto.rs]: Makes sense.  Fixed.

**Comment by alexcrichton** [comment]
Thanks! Mind running rustfmt? After that I think this is good to go

**Comment by jasondavies** [comment]
I also added AES for completeness.  Note that there are currently no tests (other than assert_instr).

**Comment by gnzlbg** [comment]
@alexcrichton let's wait for the tests here :D But @jasondavies this looks really good, keep it up!

**Comment by alexcrichton** [comment]
Oh oops sorry, I misread!

**Comment by jasondavies** [comment]
Tests added.

File: coresimd/aarch64/crypto.rs [hunk 4]
```diff
@@ -0,0 +1,428 @@
...
+    #[simd_test = "crypto"]
+    unsafe fn test_vaeseq_u8() {
+        let data = u8x16::new(1, 2, 3, 4, 5, 6, 7, 8, 1, 2, 3, 4, 5, 6, 7, 8)
+            .into_bits();
+        let key = u8x16::new(0, 1, 2, 3, 4, 5, 6, 7, 0, 1, 2, 3, 4, 5, 6, 7)
+            .into_bits();
+        let r: u8x16 = vaeseq_u8(data, key).into_bits();
+        assert_eq!(
+            r,
+            u8x16::new(
```
Comment (gnzlbg) [inline coresimd/aarch64/crypto.rs]: nitpick: no need to do this, but if you want to manually format these next time just put them in a variable and tell rustfmt to skip these:

```rust
#[cfg_attr(rustfmt, skip)]
let r = u8x16::new(
    1, 2, 3, 4, 5, 6, 7, 8
    9, 10, 11, 12, 13, 14, 15
);
```

**Comment by gnzlbg** [comment]
So looks good. I've restarted the SDE build bot in case the failure was spurious, and the rustfmt build bot is failing (maybe you need to `rustup update` to use the latest rustfmt?). Once these two are green this can be merged.

**Comment by jasondavies** [comment]
Thanks, I did notice that but wasn't sure if I should be changing so many files just to satisfy rustfmt for this PR.  Just pushed a commit that should fix it, after running `cargo fmt`.

**Comment by alexcrichton** [comment]
Thanks! I think that recent nightlies have broken CI here for i128, mind removing that feature listed while you're at it?

**Comment by alexcrichton** [comment]
Hm a new test is failing, I wonder if that's related to the formatting changes by accident?

**Comment by jasondavies** [comment]
Maybe I should revert the rustfmt commit and see if it passes?

**Comment by alexcrichton** [comment]
Sure!

**Comment by gnzlbg** [comment]
Which test is failing? The Intel SDE was already failing before the formatting changes, and AFAICT this PR only touches AArch64 anyways :/

**Comment by jasondavies** [comment]
Yeah it was just the SDE one.  I guess I should just kill the last revert commit in that case and this should be ready to merge.

**Comment by jasondavies** [comment]
The formatting changes touch a whole bunch of files but I agree it would be a bit worrying if formatting changes caused any test failures.

**Comment by jasondavies** [comment]
OK, done.

**Comment by gnzlbg** [comment]
This is the build bot where SDE is running without the formatting changes: https://travis-ci.org/rust-lang-nursery/stdsimd/jobs/358879191

If that still fails, let's just merge this and fix SDE afterwards. Maybe something on nightly has changed that breaks SDE, or maybe we just need to update SDE to a newer version.

**Comment by jasondavies** [comment]
Looks like the SDE build with no formatting changes failed.

**Comment by gnzlbg** [comment]
The last PR was 4 days ago and back then everything was green so something in the last couple of nightlies might have broken it :/

**Pull request merged**